File: community-puzzles/ancestors-and-descendants/ancestors-and-descendants.rb
def print_tree(tree)
  output = ""
  if (!tree.empty?)
    i = 0
    while (i < tree.size - 1)
      output += tree[i] + " > "
      i += 1
    end
    output += tree[i]
    puts(output)
  end
end

nb_people = gets.to_i
tree = [] # partial family tree

nb_people.times {
  line = gets.strip
  name = line.tr(".", "").chomp
  name_level = line.size - name.size
  tree_level = tree.size
  if (name_level == 0)
    if (!tree.empty?)
      print_tree(tree)
      tree.clear()
    end
  elsif (name_level < tree_level)
    print_tree(tree)
    while (name_level < tree.size)
      tree.pop()
    end
  end
  tree.push(name)
}
print_tree(tree)

File: community-puzzles/ancestors-and-descendants/ancestors_and_descendants.py
def print_tree(tree):
    output = ""
    if not tree:
        return
    i = 0
    while i < len(tree) - 1:
        output += tree[i] + " > "
        i += 1
    output += tree[i]
    print(output)


nb_people = int(input())
tree = []  # partial family tree

for _ in range(nb_people):
    line = input().strip()
    name = line.replace(".", "").rstrip()
    name_level = len(line) - len(name)
    tree_level = len(tree)
    
    if name_level == 0:
        if tree:
            print_tree(tree)
            tree.clear()
    elif name_level < tree_level:
        print_tree(tree)
        while name_level < len(tree):
            tree.pop()
    
    tree.append(name)

print_tree(tree)
